### PR TITLE
Third phase

### PR DESCRIPTION
**Pull request by fairwinds0099**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -119,6 +119,21 @@
             <artifactId>commons-collections</artifactId>
             <version>3.2.2</version>
         </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
```
Comment (fairwinds0099) [inline pom.xml]: what is Jackson? Is there an alternative?

File: pom.xml [hunk 2]
```diff
@@ -119,6 +119,21 @@
...
+            <version>2.9.7</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-core</artifactId>
+            <version>2.9.7</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
```
Comment (fairwinds0099) [inline pom.xml]: which Jackson Annotations have u used?

File: src/main/java/domain/pages/DirectoryPage.java [hunk 3]
```diff
@@ -20,6 +23,10 @@
 public class DirectoryPage extends HomePage {
 
     private String stateTextXpath = "//*[@id='%s']/..//td[2]";
+    private String districtXpath = "//*[@id='%s']/..//td[1]";
+    private String partyXpath = "//*[@id='%s']/..//td[3]";
+    private String officeRoomXpath = "//*[@id='%s']/..//td[4]";
```
Comment (fairwinds0099) [inline src/main/java/domain/pages/DirectoryPage.java]: why the xpath at line 30 not work?

File: src/main/java/tests/uiTests/leadershipTests.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        String expectedResult = "Majority Leader\n"+
+        "Rep. Kevin McCarthy\n"+
+        "Represents Republicans on the House floor.Majority Whip\n" +
+        "Rep. Steve Scalise\n"+
+        "Assists leadership in managing party's legislative program.Republican Conference Chairman\n"+
+        "Rep. Cathy McMorris Rodgers\n"+
+        "Heads organization of all Republican Party members in the House.Republican Policy Committee Chairman\n" +
+        "Rep. Luke Messer\n"+
+        "Heads Conference forum for policy development.";
+        LeadershipPage leadershipPage = new PageFactory().initElements(driver, LeadershipPage.class);
```
Comment (fairwinds0099) [inline src/main/java/tests/uiTests/leadershipTests.java]: move to super(parent) class

File: src/main/java/domain/pages/DirectoryPage.java [hunk 5]
```diff
@@ -61,6 +68,44 @@ public void clickListByStateAndDistrictLink(){
...
+        String officeRoomXpathFormatted = String.format(officeRoomXpath, state);
+        List<String> officeRooms = new ArrayList<String>();
+        List<WebElement> offices = driver.findElements(By.xpath(officeRoomXpathFormatted));
+        for (WebElement  office : offices) {
+            officeRooms.add(office.getText());
+        }
+        return officeRooms;
+    }
+
+    public String repInfoToJson(Representatives obj) throws JsonProcessingException {
```
Comment (fairwinds0099) [inline src/main/java/domain/pages/DirectoryPage.java]: rename method >>> serializeRepInfoWithJackson

File: src/main/java/domain/pages/DirectoryPage.java [hunk 6]
```diff
@@ -61,6 +68,44 @@ public void clickListByStateAndDistrictLink(){
...
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.enable(SerializationFeature.INDENT_OUTPUT);
+        String json = mapper.writeValueAsString(obj);
```
Comment (fairwinds0099) [inline src/main/java/domain/pages/DirectoryPage.java]: return mapper.writeValueAsString(obj)

File: src/main/java/tests/uiTests/RepresentativesTests.java [hunk 7]
```diff
@@ -28,6 +39,44 @@
...
+        firstAlabamaRep.setName(directoryPage.getAllRepresentativeNamesForSelectedState("state-alabama").get(0));
+        firstAlabamaRep.setParty(directoryPage.getPartysForSelectedState("state-alabama").get(0));
+        firstAlabamaRep.setState("alabama");
+        firstAlabamaRep.setOfficeRoom(directoryPage.getRepsOfficeRooms("state-alabama").get(0));
+        Assert.assertEquals("Byrne, Bradley", firstAlabamaRep.getName());
+        Assert.assertEquals("R", firstAlabamaRep.getParty());
+        Assert.assertEquals("1st", firstAlabamaRep.getDistrict());
+        Assert.assertEquals("alabama", firstAlabamaRep.getState());
+        Assert.assertEquals("119 CHOB", firstAlabamaRep.getOfficeRoom());
+        Assert.assertEquals("{\n" +
```
[the remark this answers is not in the document]
Comment (jacobbrendall) [reply src/main/java/tests/uiTests/RepresentativesTests.java]: yapistirinca kendiliginden cikti

[the remark this answers is not in the document]
Comment (jacobbrendall) [reply src/main/java/tests/uiTests/RepresentativesTests.java]: ama "" icine yapistirinca cikiyor

File: pom.xml (hunk 1 above)
Comment (jacobbrendall) [reply pom.xml]: Jackson is a data processing tool, but famous with generating and parsing JSON. In short, it is a library used to convert java object into json format. Alternative libraries are: 
JSON.simple, GSON, JSONP etc...

File: pom.xml (hunk 2 above)
Comment (jacobbrendall) [reply pom.xml]: did not use any. Omitted from pom.

File: src/main/java/domain/pages/LeadershipPage.java [hunk 8]
```diff
@@ -1,22 +1,49 @@
...
 
     public boolean isLeadershipTitlePresent() {
+        leadershipLink.click();
         return leadershipTitle.isDisplayed();
     }
+
+
+    public List<String> isRepublicanLeadsDisplayed() {
+        List<String> leadReps = new ArrayList<String>();
+        for (int i = 1; i < 5; i++) {
```
[the remark this answers is not in the document]
Comment (jacobbrendall) [reply src/main/java/domain/pages/LeadershipPage.java]: duzelttim abi
@FindBys ({@FindBy(xpath = "(//div[@class='row'])[4]/div[1]/p")})
    private List<WebElement> republicanReps;
   public List<String> isRepublicanLeadsDisplayed() {
        List<String> leadReps = new ArrayList<String>();
        for (WebElement r:republicanReps){
            leadReps.add(r.getText());
        }
        return leadReps;

File: src/main/java/tests/uiTests/UITestBase.java [hunk 9]
```diff
@@ -22,8 +23,10 @@
 
     @Before
     public void SystemSetUp() {
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("--headless","--disable-gpu");
```
[the remark this answers is not in the document]
Comment (jacobbrendall) [reply src/main/java/tests/uiTests/UITestBase.java]: done

File: src/main/java/tests/uiTests/leadershipTests.java (hunk 4 above)
Comment (jacobbrendall) [reply src/main/java/tests/uiTests/leadershipTests.java]: done

File: src/main/java/domain/pages/DirectoryPage.java (hunk 5 above)
Comment (jacobbrendall) [reply src/main/java/domain/pages/DirectoryPage.java]: done

File: src/main/java/domain/pages/DirectoryPage.java (hunk 6 above)
Comment (jacobbrendall) [reply src/main/java/domain/pages/DirectoryPage.java]: done, smart :)

File: pom.xml (hunk 2 above)
Comment (fairwinds0099) [reply pom.xml]: Are you sure? whre does Object Mapper class come from?
pls goto external lib and see where object mapper comes from

File: pom.xml (hunk 1 above)
Comment (fairwinds0099) [reply pom.xml]: will you please list all the alternative libraries and add to serializer class as private classes like we did for jackson and gson

File: pom.xml [hunk 10]
```diff
@@ -119,6 +119,16 @@
             <artifactId>commons-collections</artifactId>
             <version>3.2.2</version>
         </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
```
Comment (fairwinds0099) [inline pom.xml]: I think databind is th critical keyword here

File: src/main/java/domain/Pojos/PojoRepList.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+public class PojoRepList {
+    private String state;
+    private List<String> name;
+    private List<String> district;
+    private List<String> party;
+    private List<String> officeRoom;
+    private List<String> phone;
+    private List<String> committee_assignment;
+
```
Comment (fairwinds0099) [inline src/main/java/domain/Pojos/PojoRepList.java]: which approach is better do you think?
each rep row(line) in a given state as an object? or this way

File: src/main/java/domain/Pojos/Representatives.java [hunk 12]
```diff
@@ -2,18 +2,39 @@
...
+        this.district = district;
+        this.party = party;
+        this.officeRoom = officeRoom;
+        this.phone = phone;
+        this.committee_assignment = committee_assignment;
+    }
+
+    @Override
+    public String toString(){
+        return name + " " + district + " " + party + " " + officeRoom + " " + phone + " " + committee_assignment;
```
Comment (fairwinds0099) [inline src/main/java/domain/Pojos/Representatives.java]: oh my

File: src/main/java/tests/uiTests/RepresentativesTests.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (fairwinds0099) [reply src/main/java/tests/uiTests/RepresentativesTests.java]: Let's keep it here but we should not assert the whole jason... we have asssert item by item.

**Pull request merged**